src/Microsoft.AspNet.OData.Shared/Query/Expressions/TransformationBinderBase.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Interfaces;
+using Microsoft.OData.Edm;
+using Microsoft.OData.UriParser;
+
+namespace Microsoft.AspNet.OData.Query.Expressions
+{
+    internal class TransformationBinderBase : ExpressionBinderBase
+    {
+
+        internal TransformationBinderBase(ODataQuerySettings settings, IWebApiAssembliesResolver assembliesResolver, Type elementType,
+            IEdmModel model) : base(model, assembliesResolver, settings)
+        {
+            Contract.Assert(elementType != null);
+            this.SetElementType(elementType, Expression.Parameter(elementType, "$it"));
+        }
+
+        protected Type ElementType { get; private set; }
+       
+        protected ParameterExpression LambdaParameter { get; private set; }
+
+        protected bool ClassicEF { get; private set; }
+
+        /// <summary>
+        /// Gets CLR type returned from the query.
+        /// </summary>
+        public Type ResultClrType
+        {
+            get; protected set;
+        }
+
+        /// <summary>
+        /// Checks IQueryable provider for need of EF6 oprimization
+        /// </summary>
+        /// <param name="query"></param>
+        /// <returns>True if EF6 optimization are needed.</returns>
+        internal virtual bool IsClassicEF(IQueryable query)
+        {
+            var providerNS = query.Provider.GetType().Namespace;
+            return (providerNS == HandleNullPropagationOptionHelper.ObjectContextQueryProviderNamespaceEF6
+                || providerNS == HandleNullPropagationOptionHelper.EntityFrameworkQueryProviderNamespace);
+        }
+
+        protected void SetElementType(Type elementType, ParameterExpression lambdaParameter)
+        {
+            Contract.Assert(elementType == lambdaParameter.Type);
```

Comment:
If this is the assert, why not just have a SetLambdaParameter and use it to set both?  Or even have ElementType getter return this.LambdaParameter.Type ?